torch/nn/functional.py

```diff
@@ -4990,6 +4990,227 @@ def multi_head_attention_forward(
+            if not why_not_fast_path:
+                if key is value:
+                    if query is key:
+                        query = key = value = query.transpose(1, 0)
+                    else:
+                        query, key = [x.transpose(1, 0) for x in (query, key)]
+                        value = key
+                else:
+                    query, key, value = [x.transpose(1, 0) for x in (query, key, value)]
+
+                attn_output, attn_output_weights = torch._native_multi_head_attention(
+                    query,
+                    key,
+                    value,
+                    embed_dim_to_check,
+                    num_heads,
+                    in_proj_weight_,
+                    in_proj_bias_,
+                    out_proj_weight,
+                    out_proj_bias_,
+                    key_padding_mask if key_padding_mask is not None else attn_mask,
+                    need_weights,
+                    average_attn_weights,
+                    1 if key_padding_mask is not None else 0 if attn_mask is not None else None)
+                return attn_output.transpose(1, 0), attn_output_weights
+
+        any_nested = query.is_nested or key.is_nested or value.is_nested
+        assert not any_nested, ("MultiheadAttention does not support NestedTensor outside of its fast path. " +
+                                f"The fast path was not hit because {why_not_fast_path}")
+
+    return multi_head_attention_forward_impl(query,
+                                             key,
+                                             value,
+                                             embed_dim_to_check,
+                                             num_heads,
+                                             in_proj_weight,
+                                             in_proj_bias,
+                                             bias_k,
+                                             bias_v,
+                                             add_zero_attn,
+                                             dropout_p,
+                                             out_proj_weight,
+                                             out_proj_bias,
+                                             training,
+                                             key_padding_mask,
+                                             need_weights,
+                                             attn_mask,
+                                             use_separate_proj_weight,
+                                             q_proj_weight,
+                                             k_proj_weight,
+                                             v_proj_weight,
+                                             static_k,
+                                             static_v,
+                                             average_attn_weights)
+
+def multi_head_attention_forward_impl(
```

Comment:
Made function private

Reply:
What do you mean "without the forward path"?
Also, it is not because you use it as an implementation detail of a nn.Module that this has to be a new public API on torch.nn.functional.
You should either have `multi_head_attention_forward` or `multi_head_attention_forward_impl` public but not both. Also the `_impl` in the function name doesn't really make sense for a public API.